### ISSUE TITLE
[require-param]: false positive to TypeScript overload function when the next function has no jsdoc

### ISSUE THREAD

**Issue by otomad** [gajus/eslint-plugin-jsdoc#1434]
If an overload of a function does not contain some parameters, but the functions which has not jsdoc below it do contain these parameters, the jsdoc of the overloaded function will incorrectly report irrelevant parameters.

## Expected behavior

Nothing reported.

## Actual behavior

Parameters of other functions below the overloaded function (including other overloaded functions or implementation functions) will be accidentally absorbed into the jsdoc of this function, which should not happen.

## Minimal Reproduction

https://stackblitz.com/edit/stackblitz-starters-vnabrjsk?file=index.ts

## ESLint Config

```js
export default [
	tseslint.configs.base,
	{
		files: ["**/*.{js,ts}"],
		plugins: {
			jsdoc,
		},
		rules: {
			"jsdoc/require-param": "warn",
		},
	},
];
```

### The implement function which has no jsdoc, below the overload param-less function

Code:
```ts
/**
 * Test function with param.
 * @param foo - Test param.
 */
function myFunction(foo: string): void;
/**
 * Test function without param.
 */
function myFunction(): void;
function myFunction(foo?: string) {}
```

Report:
```
  6:1  warning  Missing JSDoc @param "foo" declaration  jsdoc/require-param

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.
```

After fixed:
```ts
/**
 * Test function with param.
 * @param foo - Test param.
 */
function myFunction(foo: string): void;
/**
 * Test function without param.
 * @param foo
 */
function myFunction(): void;
function myFunction(foo?: string) {}
```

Why? Where is param `foo` in the overload function `function myFunction(): void;`???

### The implement function has jsdoc
Code:
```ts
/**
 * Test function with param.
 * @param foo - Test param.
 */
function myFunction(foo: string): void;
/**
 * Test function without param.
 */
function myFunction(): void;
/**
 * Test function with param.
 * @param foo - Test param.
 */
function myFunction(foo?: string) {}
```

Nothing reported!

But the jsdoc for the implement function is useless, it almost same as the first overload, and TypeScript will not read it.

### Move the overload param-less function above other overload functions which has jsdoc
Code:
```ts
/**
 * Test function without param.
 */
function myFunction(): void;
/**
 * Test function with param.
 * @param foo - Test param.
 */
function myFunction(foo: string): void;
function myFunction(foo?: string) {}
```

Nothing reported!

### Move the overload param-less function above other overload functions which has no jsdoc
Code:
```ts
/**
 * Test function without param.
 */
function myFunction(): void;
function myFunction(foo: string): void;
function myFunction(foo?: string) {}
```

Report:
```
  1:1  warning  Missing JSDoc @param "foo" declaration  jsdoc/require-param
  1:1  warning  Missing JSDoc @param "foo" declaration  jsdoc/require-param

✖ 2 problems (0 errors, 2 warnings)
  0 errors and 2 warnings potentially fixable with the `--fix` option.
```

I got two param `foo` warnings, they come from the second overloaded function and the implementation function respectively, and have no relationship with the current overloaded function.

After fixed:
```ts
/**
 * Test function without param.
 * @param foo
 */
function myFunction(): void;
function myFunction(foo: string): void;
function myFunction(foo?: string) {}
```

## Environment

- Node version:  v23.6.1
- ESLint version  v9.32.0
- `eslint-plugin-jsdoc` version: v51.4.1

**Comment by otomad** [comment]
Is there any PR packaged instant preview like pkg.pr.new ? I can't test if the PR is suitable for my use case.

**Comment by brettz9** [comment]
I'm afraid not. You could install locally, add your code to a test in a test file and run the test (e.g., `pnpm test-index --rule=require-jsdoc --valid=-1` to test a valid test case you add at the end of the file).

**Comment by otomad** [comment]
I tested it and it almost suitable for my use cases.

But I also want to add an option `skipImOverloadedDeclarations`, which does the opposite of `skipImplementationOverloadedDeclarations`, which is used to skip documenting the implementation declaration of overloaded functions.

My current rule is to document every overload declarations, but not the implementation. This is also how VSCode displays documentation when you hover the pointer over a variable: it only shows the documentation for the overloads, not the implementation. Even if you hover the pointer over the implementation function itself, it will still show the documentation of the first overload.

Of course, you can also ignore it if it's difficult to implement, because it already suitable for my use cases now.

**Comment by otomad** [comment]
When I set the `exemptOverloadedImplementations` option to true, regardless of the `skipInterveningOverloadedDeclarations` option, it will no longer do any `require-jsdoc` checks on overloaded functions (non-overloaded functions are not affected), either overloaded declarations or implementations. Is this the correct behavior? This seems not what I expected.

**Comment by brettz9** [comment]
Also be sure that you have `TSDeclareFunction` in `contexts`.

**Comment by otomad** [comment]
I think it should be ok, but I would also like to test it tonight.
